system/include/emscripten/bind.h

```diff
@@ -587,7 +588,9 @@ namespace emscripten {
         template<typename T>
         inline T* getContext(const T& t) {
             // not a leak because this is called once per binding
-            return new T(t);
+            auto* ret = new T(t);
+            __lsan_ignore_object(ret);
```

Comment:
is this LSan function always guaranteed to be linked in when needed? Looks like it's in compiler-rt so I guess that's the case?

Reply:
I don't know if Emscripten has any special linking considerations, but the lsan interface header and its functions should always be available. The functions it defines just don't do anything if lsan isn't enabled.

Reply:
Nothing special. But if the body were in a library we don't link in by default, that would fail. But compiler-rt should be fine.

Reply:
Huh, I assumed this function needs to be guarded with `#if has_feature(leak_sanitizer) || has_feature(address_sanitizer)`, but I guess if it's always available and we target only Emscripten / Clang, then we don't need to worry about it.

Reply:
I imagine they should result in undefined symbols at link time because the including of the lsan libraries is all gates the use of the sanitizer command line flags.

See: https://github.com/emscripten-core/emscripten/blob/9a27b51321f20cb3c60bf39e5ea5e54be97798c6/tools/system_libs.py#L1661

Reply:
@sbc100 if that is indeed a problem, what would the fix be?

Reply:
Doesn't the pre-processor trick that @RReverser suggests work?

Or is the problem that the C++ file in question is not built with lsan enabled?

Reply:
By the way, something I was wondering about: if enabling ASAN automatically includes LSAN checks, why is LLVM returning `false` for `__has_feature(leak_sanitizer)` in ASAN mode? Is this an upstream bug?

Reply:
> Doesn't the pre-processor trick that @RReverser suggests work?
> 
> Or is the problem that the C++ file in question is not built with lsan enabled?

The lsan inteface is meant to be usable whether or not lsan is enabled, so ideally we would not have any linker magic that prevents that from happening. That's why I would rather not depend on the preprocessor to feature detect this.

Reply:
> By the way, something I was wondering about: if enabling ASAN automatically includes LSAN checks, why is LLVM returning `false` for `__has_feature(leak_sanitizer)` in ASAN mode? Is this an upstream bug?

I don't know. Are you seeing this behavior in an Emscripten test I can look at or is it in your own code?

Reply:
> The lsan inteface is meant to be usable whether or not lsan is enabled, so ideally we would not have any linker magic that prevents that from happening. That's why I would rather not depend on the preprocessor to feature detect this.

FWIW the `__has_feature(...)` condition is something we document on https://emscripten.org/docs/debugging/Sanitizers.html (same section as the `namespace __lsan {...}`).

If I understand you correctly, it's not necessary with the new stable API anymore and the interface is always available? If so, then it's worth updating those docs as part of #11644.

Reply:
> I don't know. Are you seeing this behavior in an Emscripten test I can look at or is it in your own code?

That's something in core LLVM / Clang: https://godbolt.org/z/x9MexY. So, I know it's coming from the upstream, but I'm not sure if it's a bug or, for some reason, an intended behaviour.

Reply:
I guess you could reference it via a weak reference.. but weak references don't pull things out of libraries so I guess that wouldn't work either. 

You certainly cant include a strong reference to it unless you know you are compiling with lsan.

For example on my desktop clang:

```
$ cat test.c 
#include <sanitizer/lsan_interface.h>

void* a;
int main() {
  __lsan_ignore_object(a);
  return 0;
}
$ clang test.c
/usr/bin/ld: /tmp/test-83a5d6.o: in function `main':
test.c:(.text+0x18): undefined reference to `__lsan_ignore_object'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Reply:
> The lsan inteface is meant to be usable whether or not lsan is enabled

I'm not actually sure that's true, so I'll just use the preprocessor